File: analyze_schedule.py
import json
from collections import Counter, defaultdict
from datetime import datetime


def analyze_schedule_data(input_file):
    """Analisa os dados do horário para identificar padrões e inconsistências."""

    with open(input_file, "r", encoding="utf-8") as file:
        data = json.load(file)

    horarios = data.get("data", {}).get("SHorarioAluno", [])

    # Estatísticas
    total_entries = len(horarios)
    valid_entries = 0
    subjects = Counter()
    time_slots = Counter()
    locations = Counter()
    days_of_week = Counter()

    # Mapear dias da semana
    days_map = {
        "0": "Domingo",
        "1": "Segunda",
        "2": "Terça",
        "3": "Quarta",
        "4": "Quinta",
        "5": "Sexta",
        "6": "Sábado",
    }

    # Análise por mês
    monthly_distribution = defaultdict(int)

    print("🔍 ANÁLISE DOS DADOS DO HORÁRIO ACADÊMICO")
    print("=" * 50)

    for entry in horarios:
        if entry.get("NOME") and entry.get("DATAINICIAL"):
            valid_entries += 1

            # Contadores
            subjects[entry.get("NOME")] += 1
            time_slots[f"{entry.get('HORAINICIAL')} - {entry.get('HORAFINAL')}"] += 1

            # Localização
            if entry.get("PREDIO"):
                locations[entry.get("PREDIO")] += 1

            # Dia da semana
            day = entry.get("DIASEMANA", "")
            if day in days_map:
                days_of_week[days_map[day]] += 1

            # Distribuição mensal
            try:
                date_obj = datetime.fromisoformat(
                    entry.get("DATAINICIAL").replace("T00:00:00", "")
                )
                month_year = date_obj.strftime("%Y-%m")
                monthly_distribution[month_year] += 1
            except (ValueError, AttributeError, TypeError):
                pass

    # Relatório
    print("📊 ESTATÍSTICAS GERAIS:")
    print(f"Total de registros: {total_entries}")
    print(f"Registros válidos: {valid_entries}")
    print(f"Registros inválidos: {total_entries - valid_entries}")
    print()

    print(f"📚 MATÉRIAS ({len(subjects)} diferentes):")
    for subject, count in subjects.most_common():
        print(f"   {subject}: {count} aulas")
    print()

    print("⏰ HORÁRIOS MAIS COMUNS:")
    for time_slot, count in time_slots.most_common(10):
        print(f"   {time_slot}: {count} aulas")
    print()

    print("📍 LOCAIS MAIS UTILIZADOS:")
    for location, count in locations.most_common(5):
        print(f"   {location}: {count} aulas")
    print()

    print("📅 DISTRIBUIÇÃO POR DIA DA SEMANA:")
    for day in ["Segunda", "Terça", "Quarta", "Quinta", "Sexta", "Sábado", "Domingo"]:
        count = days_of_week.get(day, 0)
        if count > 0:
            print(f"   {day}: {count} aulas")
    print()

    print("📆 DISTRIBUIÇÃO MENSAL:")
    for month in sorted(monthly_distribution.keys()):
        print(f"   {month}: {monthly_distribution[month]} aulas")
    print()


if __name__ == "__main__":
    try:
        analyze_schedule_data("data/QuadroHorarioAluno.json")
    except FileNotFoundError:
        print("❌ Erro: Arquivo 'data/QuadroHorarioAluno.json' não encontrado.")
    except Exception as e:
        print(f"❌ Erro: {e}")

File: server/src/services/analyzeSchedule.ts
import {
  ApiValidationError,
  type AnalysisResult,
  type ScheduleData,
  type ScheduleEntry,
} from "../types.js";

const DAYS_MAP: Record<string, string> = {
  "0": "Domingo",
  "1": "Segunda",
  "2": "Terça",
  "3": "Quarta",
  "4": "Quinta",
  "5": "Sexta",
  "6": "Sábado",
};

const DAY_ORDER = [
  "Segunda",
  "Terça",
  "Quarta",
  "Quinta",
  "Sexta",
  "Sábado",
  "Domingo",
] as const;

function incrementCounter(counter: Record<string, number>, key: string): void {
  counter[key] = (counter[key] ?? 0) + 1;
}

function sortCounterEntries(
  counter: Record<string, number>,
  limit?: number,
): Record<string, number> {
  const entries = Object.entries(counter).sort((left, right) => {
    if (right[1] !== left[1]) {
      return right[1] - left[1];
    }

    return left[0].localeCompare(right[0], "pt-BR");
  });

  return Object.fromEntries(limit ? entries.slice(0, limit) : entries);
}

function toValidScheduleEntries(payload: unknown): ScheduleEntry[] {
  if (!payload || typeof payload !== "object" || !("data" in payload)) {
    throw new ApiValidationError("Estrutura de dados inválida: chave 'data' ausente");
  }

  const data = (payload as ScheduleData).data;
  if (!data || typeof data !== "object" || !("SHorarioAluno" in data)) {
    throw new ApiValidationError(
      "Estrutura de dados inválida: chave 'SHorarioAluno' ausente",
    );
  }

  const rawEntries = data.SHorarioAluno;
  if (!Array.isArray(rawEntries)) {
    throw new ApiValidationError(
      "Estrutura de dados inválida: 'SHorarioAluno' deve ser uma lista",
    );
  }

  return rawEntries.filter(
    (entry): entry is ScheduleEntry => Boolean(entry) && typeof entry === "object",
  );
}

function extractMonthYear(rawDate: string): string | null {
  const normalized = rawDate.replace("T00:00:00", "");
  const match = normalized.match(/^(\d{4})-(\d{2})/);
  if (!match) {
    return null;
  }

  return `${match[1]}-${match[2]}`;
}

export function analyzeScheduleDataJson(payload: unknown): AnalysisResult {
  const entries = toValidScheduleEntries(payload);
  const subjects: Record<string, number> = {};
  const timeSlots: Record<string, number> = {};
  const locations: Record<string, number> = {};
  const daysOfWeek: Record<string, number> = {};
  const monthlyDistribution: Record<string, number> = {};

  let validEntries = 0;

  for (const entry of entries) {
    const subject = typeof entry.NOME === "string" ? entry.NOME.trim() : "";
    const startDate =
      typeof entry.DATAINICIAL === "string" ? entry.DATAINICIAL.trim() : "";

    if (!subject || !startDate) {
      continue;
    }

    validEntries += 1;
    incrementCounter(subjects, subject);

    const startTime =
      typeof entry.HORAINICIAL === "string" ? entry.HORAINICIAL.trim() : "";
    const endTime =
      typeof entry.HORAFINAL === "string" ? entry.HORAFINAL.trim() : "";
    if (startTime && endTime) {
      incrementCounter(timeSlots, `${startTime} - ${endTime}`);
    }

    const location =
      typeof entry.PREDIO === "string" ? entry.PREDIO.trim() : "";
    if (location) {
      incrementCounter(locations, location);
    }

    const weekday =
      typeof entry.DIASEMANA === "string" ? entry.DIASEMANA.trim() : "";
    if (weekday && DAYS_MAP[weekday]) {
      incrementCounter(daysOfWeek, DAYS_MAP[weekday]);
    }

    const monthYear = extractMonthYear(startDate);
    if (monthYear) {
      incrementCounter(monthlyDistribution, monthYear);
    }
  }

  return {
    statistics: {
      total_entries: entries.length,
      valid_entries: validEntries,
      invalid_entries: entries.length - validEntries,
    },
    subjects: sortCounterEntries(subjects),
    time_slots: sortCounterEntries(timeSlots, 10),
    locations: sortCounterEntries(locations, 5),
    days_of_week: Object.fromEntries(
      DAY_ORDER.filter((day) => (daysOfWeek[day] ?? 0) > 0).map((day) => [
        day,
        daysOfWeek[day] ?? 0,
      ]),
    ),
    monthly_distribution: Object.fromEntries(
      Object.entries(monthlyDistribution).sort((left, right) =>
        left[0].localeCompare(right[0]),
      ),
    ),
  };
}
